File: backtrack/code/subset-dfs-leet78.java
import java.util.*;

// leet ac 2023.11.04

class Solution {
    // cur: 将要处理第 cur 个数。从 0 开始算起。
    // 若 cur == n（最大下标 +1），说明所有数都已处理完了，可返回（cur 也不用处理了）
    void dfs(int cur, int[] nums, List<Integer> t, List<List<Integer>> res) {
        if (cur == nums.length) {
            res.add(new ArrayList<>(t));
            return;
        }
        t.add(nums[cur]);
        dfs(cur + 1, nums, t, res);
        t.remove(t.size() - 1);
        dfs(cur + 1, nums, t, res);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> t = new ArrayList<>();
        dfs(0, nums, t, res); // cur 从 0（第一个数）开始
        return res;
    }
    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3};
        List<List<Integer>> res = new Solution().subsets(nums);
        System.out.println(res);
    }
}

File: backtrack/code/subset-dfs-leet-78.cpp
#include <vector>
using namespace std;

// leet ac 2024.11.13

class Solution {
public:
    // cur: 将要处理第 cur 个数。从 0 开始算起。
    // 若 cur == n（最大下标 +1），说明所有数都已处理完了，可返回（cur 也不用处理了）
    // 可选项是两个：「选自己」、「不选自己」。不需要循环遍历所有可选项。
    void dfs(vector<vector<int>>& ans, vector<int>& t, int cur, const vector<int>& nums) {
        if (cur == nums.size()) {
            ans.push_back(t);
            return;
        }
        t.push_back(nums[cur]);
        dfs(ans, t, cur + 1, nums);
        t.pop_back();
        dfs(ans, t, cur + 1, nums);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> t;
        dfs(ans, t, 0, nums); // cur 从 0（第一个数）开始
        return ans;
    }
};